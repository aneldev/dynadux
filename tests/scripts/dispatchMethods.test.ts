import "jest";
import {
  createStore
} from "../../src";

interface ITodoAppState {
  logged: boolean;
  todos: ITodo[];
}

interface ITodo {
  id: string;
  label: string;
  done: boolean;
}

describe('Dynadux', () => {
  test('Dispatch methods', () => {
    let stateChanged = 0;

    const createTodoAppStore = (onChange: (state: ITodoAppState) => void) => {
      const dispatchedActions: { action: string, payload: any }[] = [];
      const store = createStore<ITodoAppState>({
        onChange,
        onDispatch: (action, payload) => dispatchedActions.push({action, payload}),
        initialState: {
          logged: false,
          todos: [],
        },
      });

      return {
        get state() {
          return store.state;
        },

        get dispatchedActions() {
          return dispatchedActions;
        },

        login: (logged: boolean): void => store.dispatchMethod('LOGIN', () => {
          return {logged};
        }),

        addTodo: (todo: ITodo): void => store.dispatchMethod('ADD_TODO', ({state: {todos}}) => {
          return {
            todos: todos.concat(todo),
          };
        }),

        removeTodo: (todoId: string): void => store.dispatchMethod('ADD_TODO', ({state: {todos}}) => {
          return {
            todos: todos.filter(todo => todo.id !== todoId),
          };
        }),

      };
    };

    const handleStateChange = (state: ITodoAppState) => {
      stateChanged++;
    };

    const todoAppStore = createTodoAppStore(handleStateChange);

    expect(todoAppStore.state).toMatchSnapshot('Initial state');

    todoAppStore.login(true);
    todoAppStore.addTodo({id: '301', label: 'Before work beers', done: false});
    todoAppStore.addTodo({id: '302', label: 'After work beers', done: false});

    expect(todoAppStore.state).toMatchSnapshot('First todos');

    todoAppStore.removeTodo('302');

    expect(todoAppStore.state).toMatchSnapshot('After remove of 302 todo');

    expect(stateChanged).toBe(4);

    expect(todoAppStore.dispatchedActions).toMatchSnapshot('Dispatched actions');
  });
});
