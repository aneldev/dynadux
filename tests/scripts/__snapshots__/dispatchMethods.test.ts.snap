// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dynadux Dispatch methods: After remove of 302 todo 1`] = `
Object {
  "logged": true,
  "todos": Array [
    Object {
      "done": false,
      "id": "301",
      "label": "Before work beers",
    },
  ],
}
`;

exports[`Dynadux Dispatch methods: Dispatched actions 1`] = `
Array [
  Object {
    "action": "LOGIN",
    "payload": undefined,
  },
  Object {
    "action": "ADD_TODO",
    "payload": undefined,
  },
  Object {
    "action": "ADD_TODO",
    "payload": undefined,
  },
  Object {
    "action": "ADD_TODO",
    "payload": undefined,
  },
]
`;

exports[`Dynadux Dispatch methods: First todos 1`] = `
Object {
  "logged": true,
  "todos": Array [
    Object {
      "done": false,
      "id": "301",
      "label": "Before work beers",
    },
    Object {
      "done": false,
      "id": "302",
      "label": "After work beers",
    },
  ],
}
`;

exports[`Dynadux Dispatch methods: Initial state 1`] = `
Object {
  "logged": false,
  "todos": Array [],
}
`;
